### ISSUE TITLE
Output formats

### ISSUE THREAD

**Issue by weikanglim** [Azure/azure-dev#2272]
# Problem

`azd` currently has alpha support for `--output json` for specific commands. In this mode, two things happen:

1. Console messages are wrapped in a `json` envelope, seen [here](https://github.com/Azure/azure-dev/blob/7846d74fcf3b608ddab6367eeffe2945682d4c52/cli/azd/pkg/contracts/envelope.go#L11). They conform to [JSONLines](https://jsonlines.org).
2. The final output is delivered in JSON.

The console message wrapping is done primarily to support tooling (such as VS, VSCode) that wants to provide a GUI experience on top of the CLI.

However, overloading `--output json` to do both means that a user that only cares about the final output in a contracted format, end up seeing less human-friendly messages. It also means that if the CLI wants to support a better integration experience with VS, VSCode, (such as delivering prompt designs), this may be harder without impacting a human user.

# Proposal

My initial proposal is that `--output json`, or maybe `--json`, simply means final output being a JSON format. An additional mode, for example `--machine`, indicates to the CLI that the user of the CLI is a computer and expects complete machine-readable output. We can then design an entire experience dedicated for this without breaking `--json` contract.

There are also other enhancements to consider, for example:
1. Guarantee that in `--json`, stdout only contains the final output and nothing else. All messages (regardless of verbosity) are guaranteed written to stderr. The command becomes a strict "data output" command. A question becomes of how/whether "action" commands like `provision` and `deploy` fit well with this model.
2. In `--machine` mode, error outputs are also contractual and can be used to react to specific error conditions.

# Call to action

Primarily, this issue should discuss the state of `--output`, whether things are working as expected or require refinement (with an initial proposal being presented above).

**Comment by ellismg** [comment]
In this proposed world, if a command fails and `--output json` is set, does this mean we just don't emit anything to `stdout` (because the command produced an error, not a result?) or do we write some object to stdout that represents the error?

My guess is that since `--machine` is not passed, we would emit a human readable error message (to stderr) and there would be no output from the command, which would exit with a non zero exit code?

**Comment by weikanglim** [comment]
> My guess is that since --machine is not passed, we would emit a human readable error message (to stderr) and there would be no output from the command, which would exit with a non zero exit code?

Correct. It would exit with non-zero exit code, and write error to stderr.
